### PR TITLE
Fixed a bug that appear when there are no water molecules in selection.

### PR DESCRIPTION
**Pull request by alejob**
Changes made in this Pull Request:
 - CHANGELOG updated.
- Improved print to screen format.
- Fixed orthography.

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by alejob** [comment]
I'm having this error in Travis CI:
```
...
Ran 4504 tests in 430.775s

OK (KNOWNFAIL=4, SKIP=11)

No output has been received in the last 10m0s, this potentially indicates a stalled
build or something wrong with the build itself.

Check the details on how to adjust your build configuration on:
https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

The build stop in one "Build Jobs" , but in my PC it builds OK.

Any idea?

**Comment by orbeckst** [comment]
We have been experiencing these stalled jobs recently and are not sure why. Restarting the job typically does it.

(Given that your travis check passed, you or someone else probably restarted it, or you submitted another commit that triggered the tests again.)

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 1]
```diff
@@ -820,7 +825,8 @@ def _selection_serial(self, universe, selection_str):
         selection = []
         for ts in universe.trajectory:
             selection.append(universe.select_atoms(selection_str))
-            print(ts.frame)
+            sys.stdout.write("\rSelecting atoms from frame: %i"%ts.frame)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Do not use `sys.stdout.write`: if you want to show progress, use our own `lib.log.ProgressMeter` class.

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 2]
```diff
@@ -415,6 +415,7 @@
 from six.moves import range, zip_longest
 
 import numpy as np
+import sys
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: You should not need `sys` for anything. For output, use [`lib.log.ProgressMeter`](http://www.mdanalysis.org/mdanalysis/documentation_pages/lib/log.html?highlight=progressmeter#MDAnalysis.lib.log.ProgressMeter) (see below comment).

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 3]
```diff
@@ -766,14 +767,18 @@ def _getOneDeltaPoint(self,universe, repInd, i ,t0, dt):
             valHH += self.lg2(np.dot(unitHHVector0,unitHHVectorp))
             valdip +=  self.lg2(np.dot(unitdipVector0,unitdipVectorp))
             n += 1
-        valOH = valOH/n
-        valHH = valHH/n
-        valdip = valdip/n
-        return (valOH,valHH,valdip)
+
+        if n == 0:
+            return (0,0,0)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Use [PEP 8](https://www.python.org/dev/peps/pep-0008/) formatting (space after comma) – applies to the whole file.

Note: If you start formatting the whole file according to PEP 8 *do this in a separate commit* so that it is clear that this commit only contains formatting. However, any new code that you're adding/changing should already be formatted to PEP 8.

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 4]
```diff
@@ -766,14 +767,18 @@ def _getOneDeltaPoint(self,universe, repInd, i ,t0, dt):
...
+        else:
+            valOH = valOH/n
+            valHH = valHH/n
+            valdip = valdip/n
+            return (valOH,valHH,valdip)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: How about leaving the `return` at the end?

```python
valOH, valHH, valdip = (valOH/n, valHHn, valdip/n) if n > 0 else (0, 0, 0)
return valOH, valHH, valdip
```

or even

```python
return  (valOH/n, valHHn, valdip/n) if n > 0 else (0, 0, 0)

```

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 5]
```diff
@@ -802,7 +802,12 @@ def _getMeanOnePoint(self,universe,selection1,selection_str,dt,totalFrames):
             valdipList.append(a[2])
             sumsdt +=  dt
             n += 1
-        return (sumDeltaOH/n,sumDeltaHH/n,sumDeltadip/n)
+
+        #if no water molecules remain in selection, there is nothing to get the mean, so n = 0.
+        if n == 0:
+            return (0,0,0)
+        else:
+            return (sumDeltaOH/n,sumDeltaHH/n,sumDeltadip/n)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: I like `return` at the end better than hidden in an `if`/`else`. But ultimately your choice.

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 6]
```diff
@@ -979,7 +985,8 @@ def _selection_serial(self,universe,selection_str):
         selection = []
         for ts in universe.trajectory:
             selection.append(universe.select_atoms(selection_str))
-            print(ts.frame)
+            sys.stdout.write("\rSelecting atoms from frame: %i"%ts.frame)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: use `ProgressMeter`

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 7]
```diff
@@ -1041,7 +1048,7 @@ def _repeatedIndex(self,selection,dt,totalFrames):
 
     def _getOneDeltaPoint(self,universe, repInd, i ,t0, dt):
         """
-        Give one point to promediate and get one point of the grapic  C_vect vs t
+        Gives one point to calculate the mean and gets one point of the grapic  C_vect vs t
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: grapic --> graphic ?? (Not sure what "graphic" is supposed to mean here.)

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 8]
```diff
@@ -1063,12 +1070,17 @@ def _getOneDeltaPoint(self,universe, repInd, i ,t0, dt):
...
             n += 1
-        valO = valO/n
-        return (valO)
+        
+        #if no water molecules remain in selection, there is nothing to get the mean, so n = 0.
+        if n == 0:                                                                                              
+            return (0,0,0)                                                                                      
+        else:
+            valO = valO/n
+            return (valO)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: No parentheses.
```python
return val0
```

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 9]
```diff
@@ -1084,7 +1096,12 @@ def _getMeanOnePoint(self,universe,selection1,selection_str,dt,totalFrames):
             valOList.append(a)
             sumsdt +=  dt
             n += 1
-        return sumDeltaO / n
+            
+        #if no water molecules remain in selection, there is nothing to get the mean, so n = 0.
+        if n == 0:
+            return (0,0,0)
+        else:
+            return (sumDeltaO/n)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: no parentheses, `return` at end?

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 10]
```diff
@@ -1102,7 +1119,8 @@ def _selection_serial(self,universe,selection_str):
         selection = []
         for ts in universe.trajectory:
             selection.append(universe.select_atoms(selection_str))
-            print(ts.frame)
+            sys.stdout.write("\rSelecting atoms from frame: %i"%ts.frame)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: use `ProgressMeter`

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 11]
```diff
@@ -1194,12 +1212,16 @@ def _getMeanOnePoint(self,universe,selection1,selection_str,wint,totalFrames):
                 continue
             sumDeltaP += a
             n += 1
-
-        return sumDeltaP/n
+        
+        #if no water molecules remain in selection, there is nothing to get the mean, so n = 0.
+        if n == 0:
+            return (0,0,0)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: This is weird: isn't `sumDeltaP` a scalar? Then why would you return a tuple for `n==0`??

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 12]
```diff
@@ -1084,7 +1096,12 @@ def _getMeanOnePoint(self,universe,selection1,selection_str,dt,totalFrames):
             valOList.append(a)
             sumsdt +=  dt
             n += 1
-        return sumDeltaO / n
+            
+        #if no water molecules remain in selection, there is nothing to get the mean, so n = 0.
+        if n == 0:
+            return (0,0,0)
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Why a tuple for `n==0`?? Isn't `sumDelta0` a scalar?

File: package/MDAnalysis/analysis/waterdynamics.py [hunk 13]
```diff
@@ -1218,7 +1240,8 @@ def _selection_serial(self,universe,selection_str):
         selection = []
         for ts in universe.trajectory:
             selection.append(universe.select_atoms(selection_str))
-            print(ts.frame)
+            sys.stdout.write("\rSelecting atoms from frame: %i"%ts.frame)
+            sys.stdout.flush()
```
Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: `ProgressMeter` instead

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 1 above)
Comment (kain88-de) [inline package/MDAnalysis/analysis/waterdynamics.py]: Or a good old print.

Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: ProgressMeter is made for this use case. Bare prints are not good because you cannot switch them off easily (short of redirecting stdout).

Comment (kain88-de) [inline package/MDAnalysis/analysis/waterdynamics.py]: There should be a verbosity argument to the function. I honestly never look at the log and I'm not sure how many people know about it.

Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: Maybe we misunderstand each other (?): I am not advocating to use a logger but use the `ProgressMeter` class which properly encapsulates the use case here: write status for every iteration in a loop. Under the hood it uses `sys.stderr` and flushes but with added goodies like #928 / #960 (and some other neat things that @jbarnoud has shown or promised (IIRC... ;-) ) such as real progress bars, including nice ones in Jupyter notebooks) it makes sense to unify the way we show progress.

Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: And yes, add the standard `verbose=True` argument and pass it to `ProgressMeter`.

**Comment by alejob** [comment]
Thanks for all your comments, I'll work on this.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 4 above)
Comment (alejob) [inline package/MDAnalysis/analysis/waterdynamics.py]: Ok, I like it.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 12 above)
Comment (alejob) [inline package/MDAnalysis/analysis/waterdynamics.py]: You're right, my mistake, next time I'll implement tests for this "little" changes.

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 3 above)
Comment (alejob) [inline package/MDAnalysis/analysis/waterdynamics.py]: Ok, I'll adapt the code to the PEP8 format, but is it necessary to have lines shorter than 80 column even if I lost readability of code?

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 7 above)
Comment (alejob) [inline package/MDAnalysis/analysis/waterdynamics.py]: I've changed it for "plot".

File: package/MDAnalysis/analysis/waterdynamics.py (hunk 3 above)
Comment (jbarnoud) [inline package/MDAnalysis/analysis/waterdynamics.py]: Try to keep a limit at 80-ish columns. There is no point at cutting lines at 79 for the sake of it, you 82 or 85 as just as readable, but long lines are often the sign of other issues.

Comment (orbeckst) [inline package/MDAnalysis/analysis/waterdynamics.py]: We decided to go with 79 characters, see [Style Guide: Coding Style](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#coding-style) – there was quite a bit of discussion preceding this decision. Generally speaking, I find that most code still looks fine when broken sensibly across lines.

A lot of old code still does not conform to PEP 8 but we are slowly working to making this more uniform so at least newly added/edited code should conform.

In the end, a unified appearance makes it easier for other people to understand the code and it can make merging easier, too.

**Comment by orbeckst** [comment]
@alejob all looking good, nice work! However, I wanted to consolidate your commits so I created PR #1323 . I am closing this PR and we continue at the other one.
